### DISCUSSION TITLE
Two banks affecting CHANGE_CHANNEL and CHANGE_ADDRESS

### DISCUSSION THREAD

**Discussion by birgy** [tttapa/Control-Surface#381] [Q&A]
Many thanks for your library, it is so helpful, especially for someone who does not master code writing. I have a question, though, relative to the creation of banks. My project requires that I should be able to change  the midi channel of my ccpotentiometers and ccbuttons manually (an incrementSelectorLEDS button seemed the most obvious candidate)  AND also change their addresses (CC controller number) with another incrementSelectorLEDS.
The comments in bank.ino say that I can change CHANGE_CHANNEL into CHANGE_ADDRESS, but I have found no reference to the possibility of using the two bank types to modify the address and channel of each controller.
This is the file I am using as a test:

```cpp
#include <Control_Surface.h>
auto &serial = Serial;
SerialMIDI_Interface<decltype(serial)> midi = {serial, MIDI_BAUD};

Bank<3> bank1(5); // 3 modes which offset the pots  by +5
IncrementSelectorLEDs<3> selector1 = {
  Bank,
  1,
  {8, 9, 10},  // three pins for the leds
  // Wrap::Wrap does not apply here
};
Bank<3> bank2(5); // 3 modes which offset the pots  by +5
IncrementSelectorLEDs<3> selector1 = {
  Bank,
  2,
  {11, 12, 13},  // three pins for the leds
  // Wrap::Wrap does not apply here

  Bankable::CCPotentiometer potentiometer1 = {
    {bank1, BankType::CHANGE_CHANNEL}, // bank configuration
    A0, // analog pin
    {MIDI_CC::Channel_Volume, CHANNEL_1}, // address
  };
  Bankable::CCPotentiometer potentiometer2 = {
    {bank2, BankType::CHANGE_CHANNEL}, // bank configuration
    A1, // analog pin
    {MIDI_CC::Channel_Volume, CHANNEL_2}, // address
  };

  //one of the stupid things I tried: {bank1, BankType:: CHANGE_ADDRESS, CHANGE_CHANNEL}, A4, {MIDI_CC::2, CHANNEL_4}},
  //another stupid thing:
  //{{bank1, BankType::CHANGE_CHANNEL}, A3, {MIDI_CC::1, CHANNEL_3}},
  //{{bank2, BankType::CHANGE_ADDRESS}, A3, {MIDI_CC::1, CHANNEL_3}},

};
void setup() {
  Control_Surface.begin();
} void loop() {
  Control_Surface.loop();
}
```

As you see, that also means that I had to declare more than one bank and I thought that I could freely name them  bank1 and bank2 (in a particularly eccentric flight of my imagination). But maybe I am deceived (there again I could not find an example of multiple banks using the library.)
If I am mistaken and some of the issues pages contain that information, please ignore my query. But if the information may be found elsewhere, would you have the kindness to redirect me? Thanks a lot.
Philippe

**Comment by EvgeniySavkin** [comment]
Try reading the post #308.

**Reply by birgy** [reply]
Thank you for the pointer, indeed, that seems to solve the bank problem. As for the possibility of making the emitting channel of a given potentiometer AND its control number two variables controllable by two buttons or encoders, I have searched the forum and I am afraid the answer to that one is a no...
Cheers

**Comment by tttapa** [comment]
It's possible but requires a bit of extra boilerplate.

The way most bankable elements are implemented is as follows:
 - The actual logic for reading inputs etc. is implemented in a generic class, such as `Bankable::MIDIFilteredAnalog<BankAddress, Sender>`.
 - The first template argument `BankAddress` is used to determine the address of the messages that are sent, based on the setting of a bank.
 - The second template argument `Sender` determines what type of MIDI messages to send. E.g. `Bankable::CCPotentiometer` is a `Bankable::MIDIFilteredAnalog<..., ContinuousCCSender>`, while `Bankable::PBPotentiometer` is a `Bankable::MIDIFilteredAnalog<..., PitchBendSender>`.

To get a MIDI element that has its address controlled by two banks, you can use the `SingleAddressMultipleBanks` class. You then have to compose the generic MIDI element class with this bank address class.

As an example, let's say you want something like `Bankable::CCPotentiometer`, but controlled by two banks. Start by going to the [`Bankable::CCPotentiometer` documentation](https://tttapa.github.io/Control-Surface-doc/new-input/Doxygen/d7/dc0/classBankable_1_1CCPotentiometer.html), and click on “Inheritance diagram for CCPotentiometer”. You'll see that `CCPotentiometer` inherits from `MIDIFilteredAnalog<SingleAddress, ContinuousCCSender>`. `SingleAddress` is a class that uses one bank to calculate a single address. Instead, you want a class that uses _two_ banks to compute a single address, so you use `SingleAddressMultipleBanks<2>` (`2` is the number of banks).

You would implement it as follows:

```cpp
class DualBankCCPotentiometer : public Bankable::MIDIFilteredAnalog<Bankable::SingleAddressMultipleBanks<2>, ContinuousCCSender> {
  public:
    DualBankCCPotentiometer(const Array<OutputBankableMIDIAddress, 2> &banks, pin_t pin, MIDIAddress address)
        : Bankable::MIDIFilteredAnalog<Bankable::SingleAddressMultipleBanks<2>, ContinuousCCSender>({banks, address}, pin, {}) {}
};
```
You can look at the [implementation of `Bankable::CCPotentiometer`](https://tttapa.github.io/Control-Surface-doc/new-input/Doxygen/d6/ddd/Bankable_2CCPotentiometer_8hpp_source.html) for inspiration, they are almost identical.

To use it in your sketch, you can then use:

```cpp
DualBankCCPotentiometer pot {
    {{
        {channelbank, BankType::CHANGE_CHANNEL},
        {addressbank, BankType::CHANGE_ADDRESS},
    }},                // Banks and types
    A0,                // analog input pin
    {0x07, CHANNEL_1}, // base address
};
```

Here's the full sketch:

```cpp
#include <Control_Surface.h> // Include the Control Surface library

USBDebugMIDI_Interface midi; // Debug output for easy testing

class DualBankCCPotentiometer : public Bankable::MIDIFilteredAnalog<Bankable::SingleAddressMultipleBanks<2>, ContinuousCCSender> {
  public:
    DualBankCCPotentiometer(const Array<OutputBankableMIDIAddress, 2> &banks, pin_t pin, MIDIAddress address)
        : Bankable::MIDIFilteredAnalog<Bankable::SingleAddressMultipleBanks<2>, ContinuousCCSender>({banks, address}, pin, {}) {}
};

Bank<3> channelbank(1); // 3 modes which offset the channel by +1
Bank<3> addressbank(5); // 3 modes which offset the address by +5

IncrementSelector<3> selectors[] {
  {channelbank, 2},
  {addressbank, 3},
};

DualBankCCPotentiometer pot {
    {{
        {channelbank, BankType::CHANGE_CHANNEL},
        {addressbank, BankType::CHANGE_ADDRESS},
    }},                // Banks and types
    A0,                // analog input pin
    {0x07, CHANNEL_1}, // base address
};

void setup() {
  Control_Surface.begin();
}

void loop() {
  Control_Surface.loop();
}
```

Note that I tested it using the `new-input` branch, not the `master` branch. If it doesn't work for you, please post the full error message and the version of Control Surface you're using.

**Reply by birgy** [reply]
Pieter,
many thanks for taking the time to answer, I am glad to know that this can be done within the library.
I updated the library on my IDE using the new-input branch. 

I have tried with and without resistors on the two buttons (I guess in your library, the internal pullup resistor is systemically activated), and the results are unexpected in each case.
The pot values are correctly red and transmitted but the channel numbers appear to change cyclically (following a sequence of three increments which offsets the channel by 1) while the control number is offset by 5, three times again (as in the Bank<3> addressbank(5) settings, only it does not react as expected). 

Control Change  	Channel: 1	Data 1: 0x0c	Data 2: 0x3e	Cable: 1
Control Change  	Channel: 1	Data 1: 0x0c	Data 2: 0x3f	Cable: 1
Control Change  	Channel: 1	Data 1: 0x0c	Data 2: 0x40	Cable: 1
Control Change  	Channel: 1	Data 1: 0x0c	Data 2: 0x41	Cable: 1
Control Change  	Channel: 1	Data 1: 0x0c	Data 2: 0x42	Cable: 1
Control Change  	Channel: 1	Data 1: 0x0c	Data 2: 0x43	Cable: 1
Control Change  	Channel: 2	Data 1: 0x11	Data 2: 0x44	Cable: 1
Control Change  	Channel: 2	Data 1: 0x11	Data 2: 0x45	Cable: 1
Control Change  	Channel: 2	Data 1: 0x11	Data 2: 0x46	Cable: 1
Control Change  	Channel: 2	Data 1: 0x11	Data 2: 0x47	Cable: 1
Control Change  	Channel: 2	Data 1: 0x11	Data 2: 0x48	Cable: 1
Control Change  	Channel: 3	Data 1: 0x07	Data 2: 0x49	Cable: 1
Control Change  	Channel: 3	Data 1: 0x07	Data 2: 0x4a	Cable: 1
Control Change  	Channel: 3	Data 1: 0x07	Data 2: 0x4b	Cable: 1
Control Change  	Channel: 3	Data 1: 0x07	Data 2: 0x4c	Cable: 1
Control Change  	Channel: 3	Data 1: 0x07	Data 2: 0x4d	Cable: 1
Control Change  	Channel: 3	Data 1: 0x07	Data 2: 0x4e	Cable: 1
Control Change  	Channel: 1	Data 1: 0x0c	Data 2: 0x4f	Cable: 1
Control Change  	Channel: 1	Data 1: 0x0c	Data 2: 0x50	Cable: 1
Control Change  	Channel: 1	Data 1: 0x0c	Data 2: 0x51	Cable: 1
Control Change  	Channel: 1	Data 1: 0x0c	Data 2: 0x52	Cable: 1
Control Change  	Channel: 1	Data 1: 0x0c	Data 2: 0x53	Cable: 1
Control Change  	Channel: 2	Data 1: 0x0c	Data 2: 0x54	Cable: 1
Control Change  	Channel: 2	Data 1: 0x11	Data 2: 0x55	Cable: 1
Control Change  	Channel: 2	Data 1: 0x11	Data 2: 0x56	Cable: 1
Control Change  	Channel: 3	Data 1: 0x07	Data 2: 0x57	Cable: 1
Control Change  	Channel: 3	Data 1: 0x07	Data 2: 0x58	Cable: 1
Control Change  	Channel: 3	Data 1: 0x07	Data 2: 0x59	Cable: 1
Control Change  	Channel: 3	Data 1: 0x07	Data 2: 0x5a	Cable: 1 

Philippe

**Reply by tttapa** [reply]
It's not entirely clear what you mean. The channels and addresses changing cyclically is the correct behavior when using the `IncrementSelector`: each time you press the selector button, it switches to the next selection, and when you reach the last selection, it wraps around to the first again. What does seem strange in the output you posted is that the channel and the address change at the same time. Using the code I posted, you should be able to control them individually.

All buttons in the Control Surface library automatically enable the internal pull-up resistor, so you can just connect the push buttons between ground and the input pin.  
Note that I changed the pin numbers of the selectors to be able to test it with the switches I had wired up on my breadboard, you may have to change them again to match your setup.

Edit: Ah, now I see the problem, do you mean that the channel numbers and addresses change cyclically all the time without doing anything? In that case, there's probably an issue with how the selector switches are wired, and the library reads them as “pressed” all the time. Make sure that you are using momentary, normally open switches, connected between ground and an input pin. If you hold down a selector button it works similar to the volume rocker on a phone or a keyboard key: if you press them it increments once, then there's some delay, and after that it keeps on incrementing quickly as long as you keep the button pressed.

**Reply by birgy** [reply]
Yes I was not clear, these changes occur when I move the shaft of the potentiometer, without touching the buttons. But when I push the buttons, they do not affect this behaviour... When I do not touch anything, no message is emitted. (I have checked the pin numbers for the buttons, it is ok). I am using an arduino uno rev 3.

**Reply by tttapa** [reply]
Which pins are you using for the buttons? How are they wired? There must be some reason for them to be read as LOW all the time, even when you're not pressing the buttons.

**Reply by birgy** [reply]
My buttons are on pin 2 and 3 connected to ground. It may be the connections were not 100 per cent reliable but now that is sorted, and the button on pin 3 behaves as expected, that is a relief, and the potentiometers sends consistent series of values on channel 1 only, BUT when I click on the second button on pin 2 (the one supposed to change channels), it sends something that reads "ntrol Change  	Channel: 1	D"  in the serial monitor after which everything is stalled and no message is emitted anymore.  I have tried other pins (changing the code accordingly) to check whether one of the pins was not fried on the arduino) but the problem remains the same
I am afraid that may come from poor unsoldered connections on the breadboard and I will test everything before I bother you any more, because you have been so kind to propose a solution and I do not want to waste your time.
Many thanks again

**Reply by tttapa** [reply]
That's odd. If you post the full code that triggers this issue, I'll see if I can reproduce it.

**Reply by birgy** [reply]
I have changed the breadboard and the push buttons, and it works perfectly well,
thanks to you I have a solid ground to built up my controller. I'll post the result when I am done.
Philippe

**Reply by birgy** [reply]
Pieter,
I just realize that the last message I thought I had posted was never included in this thread and I apologize for that. My breadboard was at fault and your code worked perfectly, so thank you very warmly for that.
Meanwhile I got on with my project but I am afraid I could not manage to create a class of bankable buttons whose address AND channel could be modified by two separate selectors. Starting from your model for the ccpotentiometers and the inherited diagram for MIDIIncrementDecrementButtons, and with the help of someone far far more competent than I am,  I came up with the following code:

```ruby
class DualBankIncrementDecrementButtons
       : public Bankable::MIDIIncrementDecrementButtons<Bankable::SingleAddressMultipleBanks<2>, RelativeCCSender, DigitalNoteSender> {
    public : DualBankIncrementDecrementButtons (const Array<OutputBankableMIDIAddress, 2> &banks, pin_t incrementButtonPin, pin_t decrementButtonPin, MIDIAddress address)
             : Bankable::MIDIIncrementDecrementButtons<Bankable::SingleAddressMultipleBanks<2>, RelativeCCSender, DigitalNoteSender>(
              {banks, address},
               AH::IncrementDecrementButtons(Button(incrementButtonPin), Button(decrementButtonPin)),
               1, // multiplier
               {},
               {})
               {} // function body
                };
```

I have been using the new-input branch to test this, and the IDE tells me that the class Bankable::SingleAddressMultipleBanks has no member named getFirstActiveAddress.
My friend believes it must be something that goes wrong with the version of the library I am using.
I copy below the error message given by the IDE, in case anyone has any idea what goes wrong:

> In file included from C:\Program Files (x86)\Arduino\libraries\Control-Surface-new-input\src/MIDI_Outputs/Bankable/CCIncrementDecrementButtons.hpp:4:0,
                 from C:\Program Files (x86)\Arduino\libraries\Control-Surface-new-input\src/Control_Surface.h:55,
                 from C:\Users\garga\Documents\Arduino\control surface library essais\0-fonctionne bien\00-encoder_selector_bank_program_change7\00-encoder_selector_bank_program_change7.ino:9:
C:\Program Files (x86)\Arduino\libraries\Control-Surface-new-input\src/MIDI_Outputs/Bankable/Abstract/MIDIIncrementDecrementButtons.hpp: In instantiation of 'void CS::Bankable::MIDIIncrementDecrementButtons<BankAddress, RelativeSender, ResetSender>::update() [with BankAddress = CS::Bankable::SingleAddressMultipleBanks<2>; RelativeSender = CS::RelativeCCSender; ResetSender = CS::DigitalNoteSender]':
C:\Users\garga\Documents\Arduino\control surface library essais\0-fonctionne bien\00-encoder_selector_bank_program_change7\00-encoder_selector_bank_program_change7.ino:328:1:   required from here
C:\Program Files (x86)\Arduino\libraries\Control-Surface-new-input\src/MIDI_Outputs/Bankable/Abstract/MIDIIncrementDecrementButtons.hpp:39:41: error: 'class CS::Bankable::SingleAddressMultipleBanks<2>' has no member named 'getFirstActiveAddress'; did you mean 'getActiveAddress'?
         MIDIAddress address = addresses.getFirstActiveAddress();
                               ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
                               getActiveAddress 

NB Actually I do not need IncrementDecrementButtons, just an increment button, but as I could find nothing that goes by the name of MIDIIncrementButtons, I used what I found in the documentation.

**Reply by birgy** [reply]
At least, I have managed to make a class for simple buttons:
`class DualBankButtons : public Bankable::MIDIButton<Bankable::SingleAddressMultipleBanks<2>, DigitalNoteSender> {
         public : DualBankButtons (const Array<OutputBankableMIDIAddress, 2> &banks, pin_t pin, MIDIAddress address)
             : Bankable::MIDIButton<Bankable::SingleAddressMultipleBanks<2>, DigitalNoteSender>({
              banks, address,
}, pin, {}) {}
};`

**Reply by tttapa** [reply]
It's not entirely clear what you're trying to do. As the name implies, `SingleAddressMultipleBanks` only has a single address, so it only has a `getActiveAddress()` method, but no `getFirstActiveAddress()`/`getSecondActiveAddress()` methods.

You would need something like a `DualAddressesMultipleBanks` class, which doesn't yet exist, but it should be easy enough to write it by looking at the `SingleAddressMultipleBanks` and `DualAddresses` classes.
